src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp

```diff
@@ -336,41 +387,54 @@ class boss_archimonde : public CreatureScript
+            instance->DoUpdateAchievementCriteria(ACHIEVEMENT_CRITERIA_TYPE_KILL_CREATURE, me->GetEntry(), 1, me);
+
+            // Reset scheduled events
+            events.CancelEvent(EVENT_SPELL_FEAR);
+            events.CancelEvent(EVENT_SPELL_AIR_BURST);
+            events.CancelEvent(EVENT_SPELL_GRIP_OF_THE_LEGION);
+            events.CancelEvent(EVENT_SPELL_UNLEASH_SOUL_CHARGES);
+            events.CancelEvent(EVENT_SPELL_DOOMFIRE);
+            events.CancelEvent(EVENT_SPELL_FINGER_OF_DEATH);
+            events.CancelEvent(EVENT_SPELL_HAND_OF_DEATH);
+            events.CancelEvent(EVENT_SPELL_PROTECTION_OF_ELUNE);
+            events.CancelEvent(EVENT_ENRAGE);
+        }
 
         bool CanUseFingerOfDeath()
         {
-            // First we check if our current victim is in melee range or not.
-            Unit* victim = me->GetVictim();
-            if (victim && me->IsWithinDistInMap(victim, me->GetAggroRange(victim)))
-                return false;
-
-            ThreatContainer::StorageType const &threatlist = me->getThreatManager().getThreatList();
-            if (threatlist.empty())
-                return false;
-
-            std::list<Unit*> targets;
-            ThreatContainer::StorageType::const_iterator itr = threatlist.begin();
-            for (; itr != threatlist.end(); ++itr)
+            if (me->IsAlive())
             {
-                Unit* unit = ObjectAccessor::GetUnit(*me, (*itr)->getUnitGuid());
-                if (unit && unit->IsAlive())
-                    targets.push_back(unit);
-            }
+                // First we check if our current victim is in melee range or not.
+                Unit* victim = me->GetVictim();
+                if (victim && me->IsWithinDistInMap(victim, me->GetAggroRange(victim)))
+                    return false;
 
-            if (targets.empty())
-                return false;
+                ThreatContainer::StorageType const &threatlist = me->getThreatManager().getThreatList();
+                if (threatlist.empty())
+                    return false;
 
-            targets.sort(Trinity::ObjectDistanceOrderPred(me));
-            Unit* target = targets.front();
-            if (target)
-            {
-                if (!me->IsWithinDistInMap(target, me->GetAggroRange(target)))
-                    return true;                                // Cast Finger of Death
-                else                                            // This target is closest, he is our new tank
-                    me->AddThreat(target, me->getThreatManager().getThreat(me->GetVictim()));
-            }
+                ThreatContainer::StorageType::const_iterator itr = threatlist.begin();
+                for (; itr != threatlist.end(); ++itr)
+                {
+                    Unit* unit = ObjectAccessor::GetUnit(*me, (*itr)->getUnitGuid());
+                    if (unit && unit->IsAlive() && unit->IsCharmedOwnedByPlayerOrPlayer())
```

Comment:
```IsWithinDistInMap()``` checks for targets in meele range?

Reply:
Derp, it was `IsWithinMeleeRange()`

Reply:
You could add the `IsWithinDistInMap()` check here, then later if `targets` is empty (no targets in melee) return true to cast *Finger of Death*. This way you could avoid an extra iteration.